### PR TITLE
Hosting Command Palette: Embed the command palette on client/layout

### PR DESCRIPTION
**Pull request by sejas**
- Closes https://github.com/Automattic/dotcom-forge/issues/4572

## Proposed Changes

* In this PR we embed the WPcom `Command Palette` in `client/layout` so it will be available across all calypso pages that use that layout. It is behind a feature flag, but we will remove it in the future.

## Testing Instructions

* Access http://calypso.localhost:3000/log-in
* Press `cmd+k` on Mac or `ctrl+k` on Windows
* Observe the command palette is not available
* Access `http://calypso.localhost:3000/post/$siteSlug`
* Observe the command palette displayed is the Gutenberg version
* Access `http://calypso.localhost:3000/sites`
* Observe the command palette opens correctly with the shortcut
* Navigate to other page using the command palette, like `Open my profile` or `Open hosting configuration`
* Observe the command palette opens correctly with the shortcut
* Try in Jetpack cloud and confirm the command palette is not available

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by katinthehatsite** [comment]
I am taking a look at this PR and the changes seem to work as expected:

* I accessed the palette through the Posts > Add New and saw the Gutenberg palette instead
* I accessed the `/sites` page and saw our palette that showed up as expected
* One thing that I noticed is that when I navigate to one of the commands e.g. `Open site dashboard` and then press `cmd + k`, I am not seeing any default commands available - only the Search field. This might be intentional since this is how it looks in Gutenberg but wanted to confirm if that's the case. Here are the screenshots of what I mean:

Screenshot 2023-11-28 at 1 47 00 PM

* At the moment, all the commands are located under `client/sites-dashboard/components/wpcom-smp-commands.tsx` and I am wondering if we need to move them elsewhere since we are not limiting the commands only to `/sites` anymore?

File: client/sites-dashboard/components/wpcom-smp-commands.tsx [hunk 1]
```diff
@@ -403,6 +404,15 @@ export const useCommandsArrayWpcom = ( {
...
 			icon: addNewSiteIcon,
 		},
+		{
+			name: 'manageSites',
+			label: __( 'Manage sites' ),
+			callback: ( { close }: { close: () => void } ) => {
+				close();
+				navigate( `/sites` );
+			},
+			icon: sitesIcon,
```
Comment (katinthehatsite) [inline client/sites-dashboard/components/wpcom-smp-commands.tsx]: I added a command to go back to all sites and choose an icon that seemed to be okay but not sure if it is the best choice though

File: client/sites-dashboard/components/wpcom-smp-commands.tsx [hunk 2]
```diff
@@ -403,6 +404,15 @@ export const useCommandsArrayWpcom = ( {
 			},
 			icon: addNewSiteIcon,
 		},
+		{
```
Comment (katinthehatsite) [inline client/sites-dashboard/components/wpcom-smp-commands.tsx]: Also, not sure if we should display this as the default command when the user is outside of `/sites` in case we decided to add something as opposed to keeping only the search field

File: client/sites-dashboard/components/wpcom-smp-commands.tsx (hunk 1 above)
Comment (sejas) [reply client/sites-dashboard/components/wpcom-smp-commands.tsx]: Thanks for adding the command!

File: client/sites-dashboard/components/wpcom-smp-commands.tsx (hunk 2 above)
Comment (sejas) [reply client/sites-dashboard/components/wpcom-smp-commands.tsx]: - We will handle it https://github.com/Automattic/dotcom-forge/issues/4673

**Comment by sejas** [comment]
I added the `Command Palette` in `client/layout/index.jsx` and I want to validate if that's the right place and make sure that it doesn't cause any performance/loading issues when loading pages.

**Comment by danielbachhuber** [comment]
Does this appear on the signup flows, or site switcher?

**Comment by sejas** [comment]
> Does this appear on the signup flows, or site switcher?

In signup flows no.
In site switcher, yes. Happy to exclude it if needed.

Screenshot 2023-11-29 at 12 45 18

File: client/sites-dashboard/components/wpcom-smp-commands.tsx [hunk 3]
```diff
@@ -473,6 +474,15 @@ export const useCommandsArrayWpcom = ( {
 			},
 			icon: uploadIcon,
 		},
+		{
```
Comment (tyxla) [inline client/sites-dashboard/components/wpcom-smp-commands.tsx]: Judging by the PR title and description, I didn't expect a new command to be added in this PR. Should this be part of another PR? That would be my expectation and recommendation.

File: client/layout/index.jsx [hunk 4]
```diff
@@ -342,6 +342,9 @@ class Layout extends Component {
 				{ config.isEnabled( 'legal-updates-banner' ) && (
 					<AsyncLoad require="calypso/blocks/legal-updates-banner" placeholder={ null } />
 				) }
+				{ config.isEnabled( 'yolo/command-pallette' ) && ! isJetpackCloud() && (
```
Comment (tyxla) [inline client/layout/index.jsx]: Do we really need the `isJetpackCloud` check? Can't we just make sure not to enable (or to specifically disable) the `yolo/command-pallette` feature flag in all Jetpack Cloud environments (which should already be the case I think)? Each of the Jetpack Cloud environments has its own config files.

File: client/layout/index.jsx [hunk 5]
```diff
@@ -342,6 +342,9 @@ class Layout extends Component {
...
+					<AsyncLoad require="calypso/components/command-palette" placeholder={ null } />
```
Comment (tyxla) [inline client/layout/index.jsx:346]: Thanks for lazily loading it. It's a perfect example of a feature that is not critical and can be loaded later.

File: client/layout/index.jsx (hunk 4 above)
Comment (tyxla) [inline client/layout/index.jsx]: Note that adding it here enables the command pallette only for logged-in users. That makes sense to me, as I can't think of why we'd want to enable the command palette for logged-out users. Just wanted to confirm that y'all are aware of it and that it is intentional.

**Review by tyxla — APPROVED** [review]
Looks good to me and tests well! 👍 

I've left a couple of minor questions, but nothing big. Nice work 🙌 

I'm excited about having the command palette available throughout Calypso!  🚀 

Have we considered replacing most of the existing, if not all, of the Calypso keyboard shortcuts with that feature?

I'd also recommend introducing new commands in separate PRs, since this one has a different purpose, and a single PR is recommended to ideally solve a single problem.

File: client/layout/index.jsx (hunk 4 above)
Comment (sejas) [reply client/layout/index.jsx]: My plan when releasing the feature is removing the feature flag ~`config.isEnabled( 'yolo/command-pallette' )`~ and leaving only the `! isJetpackCloud()` check.

Comment (sejas) [reply client/layout/index.jsx]: Yes!, That's expected. Thanks for confirming. We want to display the palette only for logged in users.

File: client/sites-dashboard/components/wpcom-smp-commands.tsx (hunk 3 above)
Comment (sejas) [reply client/sites-dashboard/components/wpcom-smp-commands.tsx]: You are right, I removed this new command here https://github.com/Automattic/wp-calypso/pull/84601/commits/7ba2eaa763fd3ebc3feec4cf112d3908d6faf58a
I'll add it in a separate PR 👌

File: client/layout/index.jsx (hunk 4 above)
Comment (tyxla) [reply client/layout/index.jsx]: I understand, but my comment still applies. Instead of having the `isJetpackCloud()` check, we can still rely on feature flag checks. If we don't enable the feature flag on the Jetpack Cloud environments, we'll not need the extra `isJetpackCloud()` check. After all, that's what that check does anyway:

```
const jetpackCloudEnvironments = [
	'jetpack-cloud-development',
	'jetpack-cloud-stage',
	'jetpack-cloud-horizon',
	'jetpack-cloud-production',
];

const isJetpackCloud = (): jetpackCloudEnvironments.includes( config( 'env_id' ) );
```

Comment (tyxla) [reply client/layout/index.jsx]: In other words, feature flags aren't necessarily removed when a feature is released to production. Feature flags can remain for the time being when we want to intentionally enable/disable a feature based on the Calypso environment.

Comment (sejas) [reply client/layout/index.jsx]: Thanks for clarifying !
